DRAFT: started rewrite of security properties subsection. added some proof sketches (#57)
cfrg/draft-irtf-cfrg-generic-hybrid-kems: draft-irtf-cfrg-hybrid-kems.md

    title: "Cryptographic sponge functions"
    target: https://keccak.team/files/CSF-0.1.pdf
    date: 2011
  BJKS24:
    title: "Formal verification of the PQXDH Post-Quantum key agreement protocol for end-to-end secure messaging"
    date: 2024

>     target: https://www.usenix.org/system/files/usenixsecurity24-bhargavan.pdf

Reviewer: ```suggestion
    target: https://www.usenix.org/system/files/usenixsecurity24-bhargavan.pdf
  CAMPBELL25:
    date: 2025
    target: https://mailarchive.ietf.org/arch/msg/cfrg/qp_YxofDEl5fN6W7Xyab-juwaCc/
```

a useful framework of definitions for these binding properties. Binding for KEMs
is related to other properties for KEMs and public-key encryption, such as
robustness {{GMP22}} TODO robust encryption paper cite {{ABNM}}, and collision-freeness TODO cite mohassel.

The framework given by {{CDM23}} refers to these properties with labels
of the form X-BIND-P-Q.  The first element X is the attack model---HON, LEAK, or MAL.
P,Q means that given the value P, it is hard to produce another Q that causes Decaps to succeed.
For example, LEAK-BIND-K-PK
means that for a given shared secret (K), there is a unique encapsulation key
(PK) that could have produced it, even if all of the secrets involved are
given to the adversary after the encapsulation operation is completed (LEAK).

We treat LEAK-BIND-K-PK and LEAK-BIND-K-CT to be target binding properties.
The HON-BIND security model seems too weak for real applications---real
attacks in the LEAK model are known {{LEAK-ATTACK}}. We are not aware
of any common settings where the MAL-BIND security model is needed; thus,
LEAK-BIND seems a sensible middle ground.

Implementors SHOULD think carefully about whether MAL-BIND attacks are

> feasible for their systems or protocols.

Reviewer: SHOULD is normative language and since we haven't targeted MAL-BIND and aren't giving our reader explicitly MAL-BIND options in this document (some of the concrete instances in the other document may be) we should probably take this out

## Security Non-goals for Hybrid KEMs {#non-goals}

Considerations that were considered and not included in these designs:

Anonymity {{GMP22}}, Deniability, Obfuscation, other forms of key-robustness

> or binding {{GMP22}}, {{CDM23}}

Reviewer: todo: distinguish key binding from the things we call X-BIND-K-[CT/PK] above but haven't explained yet

useful framework of definitions for these binding properties. Binding for
KEMs is related to other properties for KEMs and public-key encryption, such
as robustness {{GMP22}} {{ABN10}}, and collision-freeness {{MOHASSEL10}}.

The framework given by {{CDM23}} refers to these properties with labels of
the form X-BIND-P-Q.  The first element X is the attack model---HON, LEAK, or
MAL.  P,Q means that given the value P, it is hard to produce another Q that
causes Decaps to succeed.  For example, LEAK-BIND-K-PK means that for a given
shared secret (K), there is a unique encapsulation key (PK) that could have
produced it, even if all of the secrets involved are given to the adversary
after the encapsulation operation is completed (LEAK).

We treat LEAK-BIND-K-PK and LEAK-BIND-K-CT to be target binding properties.
The HON-BIND security model seems too weak for real applications---real
attacks in the LEAK model are known {{BJKS24}} {{FG24}}. We are not aware
of any common settings where the MAL-BIND security model is needed; thus,
LEAK-BIND seems a sensible middle ground.

The LEAK-BIND-K-PK and LEAK-BIND-K-CT properties independently allowing using

> a KEM shared secret such that finding a colliding value with the
> encapsulation key used in its computation or the ciphertext used in its
> computation is negligible. Such properties are attractive when integrating

Reviewer: ```suggestion
a KEM shared secret such that the likelihood of finding a colliding value with the
encapsulation key used in its computation or the ciphertext used in its
computation is negligible. Such properties are attractive when integrating
```

KEMs into protocols where once protocol designers would have used
Diffie-Hellman, as they can use the smaller shared secret value alone as an

> input to a protocol key schedule for example without necessarily also needing

Reviewer: ```suggestion
input to a protocol key schedule for example, without necessarily also needing
```

> to including the much larger ciphertext or the encapsulation key to be

Reviewer: ```suggestion
to include the much larger ciphertext or the encapsulation key to be
```

protected against key confusion attacks {{FG24}} or KEM re-encapsulation
attacks {{BJKS24}}. Protocol designers may still need or want to include the
ciphertext or encapsulation key into their protocol or key schedule for other
reasons, but that can be independent of the specific properties of the KEM

> and its resulting shared secret.

Reviewer: @bwesterb @pag-crypto I added this paragraph, wdyt

Implementors should not interpret the paragraph above as absolving them
of their responsibility to carefully think through whether MAL-BIND attacks
apply in their settings.

## Security Non-goals for Hybrid KEMs {#non-goals}

> Considerations that were considered and not included in these designs:

Reviewer: ```suggestion
Security properties that were considered and not included in these designs:
```

preimage resistant (C2PRI) {{XWING}}: this allows the component KEM
encapsulation key and ciphertext to be left out from the KDF input.

#### Ciphertext Second Preimage Resistant (C2PRI) Security {#c2pri}

> Roughly, C2PRI {{XWING}} says that an adversary given an honestly-generated
> key pair (sk, pk) and the result of an *honest* Encaps(pk) - call it k, c -
> cannot find a _distinct_ c' such that Decaps(sk, c') = k. This notion has
> also been described as chosen-ciphertext resistance {{CDM23}}.

Reviewer: ```suggestion
Roughly, C2PRI {{XWING}} says that an adversary given an honestly-generated
key pair (sk, pk) and the result of an *honest* Encaps(pk) - call it k, c -
cannot find a _distinct_ c' such that Decaps(sk, c') = k. A related notion has
also been described as chosen-ciphertext resistance (CCR) {{CDM23}}. C2PRI targets preimage-resistance, whereas CCR targets collision-resistance {{CAMPBELL25}}.
```

> Roughly, C2PRI {{XWING}} says that an adversary given an honestly-generated
> key pair (sk, pk) and the result of an *honest* Encaps(pk) - call it k, c -
> cannot find a _distinct_ c' such that Decaps(sk, c') = k. A related notion has
> also been described as chosen-ciphertext resistance (CCR) {{CDM23}}. C2PRI targets preimage-resistance, whereas CCR targets collision-resistance {{CAMPBELL25}}.

Reviewer: ```suggestion
Roughly, C2PRI {{XWING}} says that an adversary given an honestly-generated
key pair (sk, pk) and the result of an *honest* Encaps(pk) - call it k, c -
cannot find a _distinct_ c' such that Decaps(sk, c') = k. A related notion has
also been described as chosen-ciphertext resistance (CCR) {{CDM23}}. C2PRI targets
preimage-resistance, whereas CCR targets collision-resistance {{CAMPBELL25}}.
```

and extend it; an RNG takes sources of noisy data and transforms them into
uniform outputs.

PRGs are related to extendable output functions (XOFs) which can be
built from random oracles. Examples include SHAKE256.

### Security Properties of PRE {#security-pre}

> The PRE hybrid KEM uses a function KeyHash to generate a short digest of the encapsulation keys.
> For IND-CCA2 of PRE, this function must be collision-resistant.

Reviewer: ```suggestion
The PRE hybrid KEM framework uses a function KeyHash to generate a short digest of the encapsulation keys.
For IND-CCA2 of PRE, this function must be collision-resistant.
```

TODO: revisit after binding properties

> ## Security Properties of Hybrid KEMs in this document

Reviewer: ```suggestion
## Security Properties of Hybrid KEMs Frameworks
```

We also argue that this extension applies, again with nearly trivial modifications, to prove security of PRE. Observe that the only difference between GHP and PRE is prehashing of the encapsulation keys. As long as the hash function is collision resistant, any event that happens in the IND-CCA2 game of GHP happens only with negligible probability in the IND-CCA2 game of PRE.

We reiterate that modulo some low-level technical details, our requirement that the KDF is indifferentiable from an RO implies that, in the ROM, the KDF used in GHP and PRE meets the split-key pseudorandomness property used in GHP's analysis.

> To summarize, all three hybrid KEMs in this document are IND-CCA2 when instantiated
> with cryptographic components that
> meet the security requirements described above. Any changes to the routines,
> including key generation/derivation, are not guaranteed to produce
> secure results.

Reviewer: ```suggestion
Therefore all three hybrid KEMs in this document are IND-CCA2 when instantiated with
cryptographic components that meet the security requirements described above. Any
changes to the algorithms, including key generation/derivation, are not guaranteed
to produce secure results.
```

### Security Requirements for PRGs {#security-prgs}

The functions used to expand a key seed to multiple key seeds is closer to a
pseudorandom generator (PRG) in its security requirements {{AOB+24}}.  A
secure PRG is an algorithm PRG : {0, 1}<sup>n</sup> → {0, 1}<sup>m</sup>,
such that no polynomial-time adversary can distinguish between PRG(r) (for r
$← {0, 1}<sup>n</sup>) and a random z $← {0, 1}<sup>m</sup> {{Rosulek}}.  The
uniform string r ∈ {0, 1}<sup>n</sup> is called the seed of the PRG.

A PRG is not to be confused with a random (or pseudorandom) _number_
generator (RNG): a PRG requires the seed randomness to be chosen uniformly
and extend it; an RNG takes sources of noisy data and transforms them into
uniform outputs.

PRGs are related to extendable output functions (XOFs) which can be
built from random oracles. Examples include SHAKE256.

### Security Properties of PRE {#security-pre}

The PRE hybrid KEM framework uses a function `KeyHash` to generate a short

> digest of the encapsulation keys.  This function must be collision-resistant.

Reviewer: ```suggestion
digest of the encapsulation keys.  This function MUST be collision-resistant.
```
